src/site_upload/check_if_complete/check_if_complete.py

```diff
@@ -0,0 +1,140 @@
+sns_client = boto3.client("sns")
+g_client = boto3.client("glue")
+
+
+def check_if_complete(message) -> (bool, dict):
+    transaction = functions.get_s3_json_as_dict(
+        bucket=os.environ.get("BUCKET_NAME"),
+        key=(
+            f"{enums.BucketPath.META.value}/transactions/{message['site']}__{message['study']}.json"
+        ),
+        s3_client=s3_client,
+    )
+    upload_time = datetime.datetime.fromisoformat(transaction["uploaded_at"])
+    for filetype, source, suffix in [
+        ("cube", enums.BucketPath.AGGREGATE.value, "aggregate"),
+        ("annotated_cube", enums.BucketPath.AGGREGATE.value, "aggregate"),
+        ("flat", enums.BucketPath.FLAT.value, "flat"),
+    ]:
+        filenames = transaction.get(filetype, [])
+        for filename in filenames:
+            dp = filename.split("__")[1].split(".")[0]
+            if source == enums.BucketPath.FLAT.value:
+                key = (
+                    f"{source}/{message['study']}/{message['site']}/{message['study']}__{dp}__{message['site']}__{transaction['version']}/"
+                    f"{message['study']}__{dp}__{message['site']}__{suffix}.parquet"
+                )
+            else:
+                key = (
+                    f"{source}/{message['study']}/{message['study']}__{dp}/{message['study']}__{dp}__{transaction['version']}/"
+                    f"{message['study']}__{dp}__{suffix}.parquet"
+                )
+            try:
+                head = s3_client.head_object(Bucket=os.environ.get("BUCKET_NAME"), Key=key)
+            except s3_client.exceptions.ClientError:
+                return False, None
+            delta = head["LastModified"] - upload_time
+            # an average time to process from an upload to all the files being
+            # processed is a minute, and since this is largely parallel it shouldn't
+            # change by much; we'll set a large buffer just in case, which should
+            # still exclude dangling files from error states
+            if upload_time > head["LastModified"] or (delta.seconds >= 300 and delta.seconds < 0):
+                return False, None
+    return True, transaction
+
+
+def has_new_packages(message, transaction) -> bool:
+    db = os.environ.get("GLUE_DB_NAME")
+    s3_bucket_name = os.environ.get("BUCKET_NAME")
+    tables = awswrangler.athena.read_sql_query(
+        (
+            f"SELECT table_name FROM information_schema.tables "  # noqa: S608
+            f"WHERE table_schema = '{db}' "  # nosec
+            f"AND regexp_like(table_name, '^{message['study']}__')"
+        ),
+        database=db,
+        s3_output=f"s3://{s3_bucket_name}/awswrangler",
+        workgroup=os.environ.get("WORKGROUP_NAME"),
+    )["table_name"].values.tolist()
+    for dp_type in ["cube", "annotated_cube", "flat"]:
+        filenames = transaction.get(dp_type, [])
+
+        for filename in filenames:
+            dp = filename.split("__")[1].split(".")[0]
+            if dp_type == "flat":
```

Comment:
This branching logic on flat feels very similar to previous code - worth centralizing probably, as these keys are so specific and easy to mess up, if you're not copying/pasting - and if you're doing that, it should be centralized :smile:

Reply:
yeah, agree - was going to do it as a followup due to the size of this.